cmd/batch-expire.go

```diff
@@ -0,0 +1,414 @@
+	Delay    time.Duration `yaml:"delay" json:"delay"`       // delay between each retries
+}
+
+// Validate validates input replicate retries.
+func (r BatchExpireRetry) Validate() error {
+	if r.Attempts < 0 {
+		return errInvalidArgument
+	}
+
+	if r.Delay < 0 {
+		return errInvalidArgument
+	}
+
+	return nil
+}
+
+// BatchExpireFilter holds all the filters currently supported for batch replication
+type BatchExpireFilter struct {
+	NewerThan     time.Duration   `yaml:"newerThan,omitempty" json:"newerThan"`
+	OlderThan     time.Duration   `yaml:"olderThan,omitempty" json:"olderThan"`
+	CreatedAfter  time.Time       `yaml:"createdAfter,omitempty" json:"createdAfter"`
+	CreatedBefore time.Time       `yaml:"createdBefore,omitempty" json:"createdBefore"`
+	Tags          []BatchExpireKV `yaml:"tags,omitempty" json:"tags"`
+	Metadata      []BatchExpireKV `yaml:"metadata,omitempty" json:"metadata"`
+}
+
+// BatchExpireNotification success or failure notification endpoint for each job attempts
+type BatchExpireNotification struct {
+	Endpoint string `yaml:"endpoint" json:"endpoint"`
+	Token    string `yaml:"token" json:"token"`
+}
+
+// BatchJobExpireFlags various configurations for replication job definition currently includes
+// - filter
+// - notify
+// - retry
+type BatchJobExpireFlags struct {
+	Filter BatchExpireFilter       `yaml:"filter" json:"filter"`
+	Notify BatchExpireNotification `yaml:"notify" json:"notify"`
+	Retry  BatchExpireRetry        `yaml:"retry" json:"retry"`
+}
+
+// BatchJobExpireV1 v1 of batch key rotation job
+type BatchJobExpireV1 struct {
+	APIVersion string              `yaml:"apiVersion" json:"apiVersion"`
+	Flags      BatchJobExpireFlags `yaml:"flags" json:"flags"`
+	Bucket     string              `yaml:"bucket" json:"bucket"`
+	Prefix     string              `yaml:"prefix" json:"prefix"`
+}
+
+// Notify notifies notification endpoint if configured regarding job failure or success.
+func (r BatchJobExpireV1) Notify(ctx context.Context, body io.Reader) error {
+	if r.Flags.Notify.Endpoint == "" {
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, r.Flags.Notify.Endpoint, body)
+	if err != nil {
+		return err
+	}
+
+	if r.Flags.Notify.Token != "" {
+		req.Header.Set("Authorization", r.Flags.Notify.Token)
+	}
+
+	clnt := http.Client{Transport: getRemoteInstanceTransport}
+	resp, err := clnt.Do(req)
+	if err != nil {
+		return err
+	}
+
+	xhttp.DrainBody(resp.Body)
+	if resp.StatusCode != http.StatusOK {
+		return errors.New(resp.Status)
+	}
+
+	return nil
+}
+
+// Expire rotates encryption key of an object
+func (r *BatchJobExpireV1) Expire(ctx context.Context, api ObjectLayer, objInfo ObjectInfo) error {
+	srcBucket := r.Bucket
+	srcObject := objInfo.Name
+
+	_, err := api.DeleteObject(ctx, srcBucket, srcObject, ObjectOptions{
+		DeletePrefix: true,
```

Comment:
allowing filters on batch expiry is incorrect imo - since a prefix delete is being done here and will apply to all versions for the matching object, though user intent is to expire specific versions that match the filter critieria. 

Either filters should be removed or this implementation should do a multi delete @harshavardhana

Reply:
No @poornas every filter is applied at latest object, it doesn't matter how many versions you have. That is not the purpose of this implementation.